### PR TITLE
feat(api-v2): Remove ForbiddenResource

### LINKED ISSUE #1543: Use ForbiddenResource consistently
If you request two resources from `/v2/resources`, and you only have permission to see one of them, you should get one resource and one `ForbiddenResource`.

Also, #1521 changed Gravsearch so that if you don't have permission to see a dependent resource, you get a link value with just the target resource's IRI. This is also what `/v2/resources` does. It would be more consistent to use `ForbiddenResource` for the dependent resource, too.

### PR DESCRIPTION
**Pull request by benjamingeer**
User story: https://dasch.myjetbrains.com/youtrack/issue/DSP-36

Discussion:

https://discuss.dasch.swiss/t/how-to-deal-with-forbidden-resources-and-values-consistently/153

https://discuss.dasch.swiss/t/knora-api-js-lib-pr-changing-public-api/158/5

This PR removes `ForbiddenResource` to make Gravsearch more consistent with the rest of API v2.

Instead of using `ForbiddenResource`, Gravsearch now returns `"knora-api:mayHaveMoreResults": true` if the number of main resources returned, plus the number of main resources hidden because of permissions, equals the page size.

This has a couple of other advantages:

- The client doesn't need to know the maximum page size (which is a server configuration setting that the client doesn't have access to) to find out if it can request another page.
- The response is more concise.

The rest of API v2 now also distinguishes between a resource not being found (`NotFoundException`, HTTP 404) and the user not having permission to see it (`ForbiddenException`, HTTP 403).

This PR also unifies the processing of search responses and other resource responses in `ConstructResponseUtilV2`.

Tasks:

- [x] Change `knora-base.ttl` and `system-data.ttl`.
- [x] Change `ConstructResponseUtilV2` and its dependencies.
- [x] Update tests and test data.
- [x] Add a client test response that includes `knora-api:mayHaveMoreResults`.
- [x] Update API docs.
- [x] Update design docs.
- [x] Increase the `knora-base` version number.

Resolves #1543.

### CODE REVIEW

**Comment by benjamingeer** [comment]
@tobiasschweizer What should happen if you request a value that you don't have permission to see. Currently you get `NotFoundException`. Do we need a `ForbiddenValue` like `ForbiddenResource`?

**Comment by tobiasschweizer** [comment]
This requires a change in knora-api-lib-js. I will create an issue. 

I don't think that the BEOL GUI needs to be adapted since it determines the max number of pages by looking at the count query result. @benjamingeer This means the this PR wouldn't break this, right?

**Comment by benjamingeer** [comment]
> This means the this PR wouldn't break this, right?

If it doesn't count the number of results in each page, and doesn't do anything with `ForbiddenResource`, it should still work. But it's probably best to test it and see. :)

**Comment by tobiasschweizer** [comment]
> > This means the this PR wouldn't break this, right?
> 
> If it doesn't count the number of results in each page, and doesn't do anything with `ForbiddenResource`, it should still work. But it's probably best to test it and see. :)

Ok, so that's a case for the staging sever. I still think this related to release management since in the end different pieces of software come together that are supposed to interact correctly, requiring compatible versions .

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 1]
```diff
@@ -1215,97 +1242,56 @@ object ConstructResponseUtilV2 {
...
+                          calculateMayHaveMoreResults: Boolean,
+                          versionDate: Option[Instant],
+                          responderManager: ActorRef,
+                          targetSchema: ApiV2Schema,
+                          settings: SettingsImpl,
+                          requestingUser: UserADM)(implicit stringFormatter: StringFormatter, timeout: Timeout, executionContext: ExecutionContext): Future[ReadResourcesSequenceV2] = {
+
+        val visibleResourceIris: Seq[IRI] = orderByResourceIri.filter(resourceIri => mainResourcesAndValueRdfData.resources.keySet.contains(resourceIri))
 
         // iterate over orderByResourceIris and construct the response in the correct order
```
Comment (SepidehAlassi) [inline webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1271]: update the comment 
iterate over `orderByResourceIris` should be iterate over `visibleResourceIris`

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 2]
```diff
@@ -1215,97 +1242,56 @@ object ConstructResponseUtilV2 {
...
-                        // include the forbidden resource instead of skipping (the amount of results should be constant -> limit)
-                        Future(forbiddenResource.getOrElse(throw AssertionException(s"Not all resources are visible, but forbiddenResource is None")))
-
-                }
         }.toVector
 
-        Future.sequence(readResourceFutures)
+        for {
+            resources <- Future.sequence(readResourceFutures)
+            mayHaveMoreResults = calculateMayHaveMoreResults && orderByResourceIri.size == settings.v2ResultsPerPage
```
Comment (SepidehAlassi) [inline webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1290]: shouldn't this be `visibleResourceIris.size == settings.v2ResultsPerPage`?

File: webapi/src/main/scala/org/knora/webapi/util/StringFormatter.scala [hunk 3]
```diff
@@ -3020,35 +3021,6 @@ class StringFormatter private(val maybeSettings: Option[SettingsImpl] = None, ma
         s"http://$IriDomain/permissions/$shortcode/$knoraPermissionUuid"
     }
 
-    /**
```
Comment (SepidehAlassi) [inline webapi/src/main/scala/org/knora/webapi/util/StringFormatter.scala:3023]: these parts seem to be not relevant to the main goal of current PR. Can they be moved to a separate PR?

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala [hunk 4]
```diff
@@ -811,10 +810,13 @@ object DeleteOrEraseResourceRequestV2 extends KnoraJsonLDRequestReaderV2[DeleteO
...
  * Represents a sequence of resources read back from Knora.
  *
- * @param numberOfResources the amount of resources returned.
- * @param resources         a sequence of resources.
+ * @param resources          a sequence of resources.
+ * @param hiddenResourceIris the IRIs of resources that were requested but that the user did not have permission to see.
+ * @param mayHaveMoreResults `true` if more resources matching the request may be available.
  */
-case class ReadResourcesSequenceV2(numberOfResources: Int, resources: Seq[ReadResourceV2]) extends KnoraResponseV2 with KnoraReadV2[ReadResourcesSequenceV2] with UpdateResultInProject {
+case class ReadResourcesSequenceV2(resources: Seq[ReadResourceV2],
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:817]: No, it's only the resources that the user is allowed to see.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 2 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1290]: No, the logic is like this: if the page was full **before** filtering for permissions, then there might be another page to request.

**Comment by benjamingeer** [comment]
@SepidehAlassi 

> I suggest that, in another branch/PR, you either remove these tests or replace them with tests to check the compatibility of releases, not the compatibility of commits, as you suggested yourself.

I actually have no idea what these tests do, how they work, or even where the code is that runs them. @tobiasschweizer and @subotic set this up and I don't know if it's documented anywhere. I could probably figure out how to turn them off, but maybe @tobiasschweizer and @subotic would be better able to deal with this properly.

> After that PR is merged to the develop branch, with a control merge, the problem of failing tests of this PR must be solved.

Does this mean I can't merge this PR until someone writes a whole new testing framework for compatibility between Knora and `knora-api-js-lib`? That sounds like it could take a long time.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 2 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1290]: I've done some further refactoring to make this clearer in #1626.

**Comment by tobiasschweizer** [comment]
> @SepidehAlassi
> 
> > I suggest that, in another branch/PR, you either remove these tests or replace them with tests to check the compatibility of releases, not the compatibility of commits, as you suggested yourself.
> 
> I actually have no idea what these tests do, how they work, or even where the code is that runs them. @tobiasschweizer and @subotic set this up and I don't know if it's documented anywhere. I could probably figure out how to turn them off, but maybe @tobiasschweizer and @subotic would be better able to deal with this properly.
> 
> > After that PR is merged to the develop branch, with a control merge, the problem of failing tests of this PR must be solved.
> 
> Does this mean I can't merge this PR until someone writes a whole new testing framework for compatibility between Knora and `knora-api-js-lib`? That sounds like it could take a long time.

I think we should have a look at this today. We could decide on a good time at the end of the Scrum meeting.

**Comment by SepidehAlassi** [comment]
@benjamingeer 
> Does this mean I can't merge this PR until someone writes a whole new testing framework for compatibility between Knora and `knora-api-js-lib`? That sounds like it could take a long time.

No, what I mean is that: Since the problems with tests should be resolved anyway before the merge, I suggest you solve the problem in another branch that can be reviewed fast, (even if solving the test problems means removing the tests) then merge it to develop first, then merge the develop branch in this branch.

**Comment by benjamingeer** [comment]
> I suggest you solve the problem in another branch that can be reviewed fast, (even if solving the test problems means removing the tests) then merge it to develop first, then merge the develop branch in this branch.

OK will do.

File: docs/src/paradox/05-internals/design/api-v2/gravsearch.md [hunk 5]
```diff
@@ -228,32 +228,39 @@ the main query can specifically ask for more detailed information on these resou
...
-All the resources and values not present in the input query's CONSTRUCT clause are filtered out. This only happens after permission checking.
-
-The main resources that have been filtered out due to insufficient permissions are represented by the placeholder `ForbiddenResource`.
-This placeholder stands for a main resource that cannot be returned, nevertheless it informs the client that such a resource exists.
-This is necessary for a consistent behaviour when doing paging.
+To do the permission checking, the results of the main query are passed to
+`ConstructResponseUtilV2.splitMainResourcesAndValueRdfData`,
+which transforms a `SparqlConstructResponse` (a set of RDF triples)
+into a structure organized by main resource Iris. In this structure, dependent
+resources and values are nested can be accessed via their main resource,
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply docs/src/paradox/05-internals/design/api-v2/gravsearch.md]: Fixed in 6e96460.

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala [hunk 6]
```diff
@@ -890,28 +905,55 @@ case class ReadResourcesSequenceV2(numberOfResources: Int, resources: Seq[ReadRe
...
      * @param requestedResourceIri the IRI of the expected resource.
      * @return the resource.
+     * @throws NotFoundException   if the resource is not found.
+     * @throws ForbiddenException  if the user does not have permission to see the requested resource.
+     * @throws BadRequestException if more than one resource was returned.
      */
-    def toResource(requestedResourceIri: IRI): ReadResourceV2 = {
-        if (numberOfResources == 0) {
-            throw AssertionException(s"Expected one resource, <$requestedResourceIri>, but no resources were returned")
+    def toResource(requestedResourceIri: IRI)(implicit stringFormatter: StringFormatter): ReadResourceV2 = {
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:916]: Always trying :)

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:817]: Added a comment in 6e96460 to make this clear.

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala [hunk 7]
```diff
@@ -890,28 +905,55 @@ case class ReadResourcesSequenceV2(numberOfResources: Int, resources: Seq[ReadRe
...
+     */
+    def checkResourceIris(targetResourceIris: Set[IRI], resourcesSequence: ReadResourcesSequenceV2): Unit = {
+        val hiddenTargetResourceIris: Set[IRI] = targetResourceIris.intersect(resourcesSequence.hiddenResourceIris)
+
+        if (hiddenTargetResourceIris.nonEmpty) {
+            throw ForbiddenException(s"You do not have permission to see one or more resources: ${hiddenTargetResourceIris.map(iri => s"<$iri>").mkString(", ")}")
         }
 
-        resourceInfo
+        val missingResourceIris: Set[IRI] = targetResourceIris -- resourcesSequence.resources.map(_.resourceIri).toSet
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:952]: Thanks! This is why I like Scala. :)

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 8]
```diff
@@ -541,41 +566,45 @@ object ConstructResponseUtilV2 {
...
+         */
         def nestResources(resourceIri: IRI, alreadyTraversed: Set[IRI] = Set.empty[IRI]): ResourceWithValueRdfData = {
-            val resource = flatResourcesWithValues(resourceIri)
+            val resource = visibleResources(resourceIri)
 
             val transformedValuePropertyAssertions: RdfPropertyValues = resource.valuePropertyAssertions.map {
                 case (propIri, values) =>
-                    val transformedValues = values.map {
-                        value =>
+                    val transformedValues: Seq[ValueRdfData] = values.foldLeft(Vector.empty[ValueRdfData]) {
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: The whole `splitMainResourcesAndValueRdfData` method has always been a headache for me to read: too long and too much nesting. Each time I have to work on it, I try to make it a bit clearer, but it really needs to be split up into much smaller pieces. I was going to ask you to refactor it with Tobias, since he's the author. :) For now, I've just moved the `nestResources` method out to the top level, and I've moved the `foldLeft` operation into another function, `transformValuesByNestingResources` (in 6e96460). But even after doing that, `splitMainResourcesAndValueRdfData` is still 273 lines long. :(

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 9]
```diff
@@ -541,41 +566,45 @@ object ConstructResponseUtilV2 {
...
+                                    // Do we have the dependent resource?
+                                    if (dependentResourceIrisVisible.contains(dependentResourceIri)) {
+                                        // Yes. Nest it in the link value.
                                         val dependentResource: ResourceWithValueRdfData = nestResources(dependentResourceIri, alreadyTraversed + resourceIri)
 
-                                        value.copy(
+                                        acc :+ value.copy(
                                             nestedResource = Some(dependentResource)
                                         )
+                                    } else if (dependentResourceIrisNotVisible.contains(dependentResourceIri)) {
```
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: > now if the dependant resource is forbidden (i.e. user does not have permission to see), then it is skipped. If the dependant resource is deleted, the link value is however returned. Have I understood it right?

Yes. Do you think this makes sense? Or should we do the same thing in both cases?

> If so, when a user requests a resource, the link value property itself will not be visible to the user either if the user does not have permission to see the dependent resource, right?

Right, it should not be visible. I will check and see if that's what really happens.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 2 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1365]: And in this PR I've just added some more comments to make this clear (in 6e96460).

File: webapi/src/main/scala/org/knora/webapi/util/StringFormatter.scala (hunk 3 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/StringFormatter.scala]: OK, I undid this change in 6e96460 and will move it to another PR. These are just functions that I forgot to delete a long time ago.

**Comment by benjamingeer** [comment]
@tobiasschweizer 

> I think we should have a look at this today. We could decide on a good time at the end of the Scrum meeting.

OK great, thanks.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 1 above)
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: Done in 6e96460, thanks.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 2 above)
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala:1365]: perfect, thanks.

**Comment by benjamingeer** [comment]
@tobiasschweizer You asked for a copy-and-paste of the failing JS lib tests in this PR:

```
**************************************************
*                    Failures                    *
**************************************************

1) workspace-project App perform a fulltext search
  - Expected '15' to equal '21'.

2) workspace-project App perform a fulltext search count query
  - Expected '15' to equal '21'.

3) workspace-project App perform a label search
  - Expected '15' to equal '21'.

4) workspace-project App perform an extended search
  - Expected '23' to equal '25'.
```

File: webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:817]: yeah, I had figured that out after reading the rest of the code yesterday. Forgot to delete this question. But thanks for adding clarifying comments.

**Comment by tobiasschweizer** [comment]
> @tobiasschweizer You asked for a copy-and-paste of the failing JS lib tests in this PR:
> 
> ```
> **************************************************
> *                    Failures                    *
> **************************************************
> 
> 1) workspace-project App perform a fulltext search
>   - Expected '15' to equal '21'.
> 
> 2) workspace-project App perform a fulltext search count query
>   - Expected '15' to equal '21'.
> 
> 3) workspace-project App perform a label search
>   - Expected '15' to equal '21'.
> 
> 4) workspace-project App perform an extended search
>   - Expected '23' to equal '25'.
> ```

Ok, this is because knora-api returns less results now. I can easily fix this in a separate knora-api-js-lib PR, so then the tests pass again in both repos. In the meantime please go ahead as agreed.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: It turns out that the link property was not filtered out if we filtered out all its values. I've fixed this and added a test in 35fafe4. Good catch, thanks!

[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: We have to return a `LinkValue` that points to a deleted resource, so the user can change the link to point to a different resource. In the case of a `LinkValue` that points to a resource that the user doesn't have permission to see, I filtered it out because @loicjaouen [said he preferred it that way](https://discuss.dasch.swiss/t/how-to-deal-with-forbidden-resources-and-values-consistently/153/2?u=benjamingeer). I think it makes sense because there isn't anything that the user could do with that link.

[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: > Yes. Do you think this makes sense? Or should we do the same thing in both cases?
yes, I believe it makes sense as it is now. In this way, the difference between a deleted dependant resource and a forbidden one is clear and the user will not be confused by seeing a link value but not its dependant resource.

[the remark this answers is not in the document]
Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: OK great. In b990663 I've added a test that gets a link to a deleted resource, to check that the link value is returned correctly.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: > For now, I've just moved the `nestResources` method out to the top level, and I've moved the `foldLeft` operation into another function, `transformValuesByNestingResources` (in [6e96460](https://github.com/dasch-swiss/knora-api/commit/6e964606936157a6198f41a7250bb86f58c6065f)). But even after doing that, `splitMainResourcesAndValueRdfData` is still 273 lines long. :(

It took me really long time to understand what the `splitMainResourcesAndValueRdfData` does, but at least now with `transformValuesByNestingResources` it is a tiny bit easier to read.

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (SepidehAlassi) [reply webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: Looks good to me!

**Comment by benjamingeer** [comment]
@SepidehAlassi Is this PR OK to merge now? If so, please don't forget to click the "approve" button.

**Comment by SepidehAlassi** [comment]
> @SepidehAlassi Is this PR OK to merge now? If so, please don't forget to click the "approve" button.

yes, I believe it is. :-)

**Comment by benjamingeer** [comment]
@SepidehAlassi Many thanks for the review!
